pkg/sqlutil/hook.go

```diff
@@ -0,0 +1,122 @@
+package sqlutil
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/jmoiron/sqlx"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+)
+
+var _ Queryer = &HookingQueryer{}
+
+// HookingQueryer is a [Queryer] which invokes a [QueryHook] on each call.
+type HookingQueryer struct {
+	q    Queryer
+	lggr logger.Logger
+	hook QueryHook
+}
+
+// QueryHook is a func that is executed for each query, providing an opportunity to measure, log, inspect/modify errors, etc.
+// The do func *must* be called.
+// Logs emitted through the provided logger.Logger will have the caller's line info.
+//
+// See [MonitorHook] and [TimeoutHook] for examples.
+type QueryHook func(ctx context.Context, lggr logger.Logger, do func(context.Context) error, query string, args ...any) error
+
+// NewHookingQueryer returns a new [HookingQueryer] that calls each [QueryHook] in the provided order.
+func NewHookingQueryer(q Queryer, l logger.Logger, hs ...QueryHook) *HookingQueryer {
+	iq := HookingQueryer{q: q,
+		lggr: logger.Helper(l, 2), // skip our own wrapper and one interceptor
+		hook: noopHook,
+	}
+	switch len(hs) {
+	case 0:
+		return &iq
+	case 1:
+		iq.hook = hs[0]
+		return &iq
+	}
+
+	for i := len(hs) - 1; i >= 0; i-- {
```

Comment:
please add a comment to guide future readers. i can't figure out what this is doing after 3 minutes of reading it.

Reply:
Will do. It is chaining them by exploiting the `do func` that they are already expected to call. The tricky part from getting the logger skip levels right.